norse/torch/functional/stdp.py

```diff
@@ -1,400 +1,276 @@
-    # ReLU to ensure weights outside [w_min, w_max] move towards it (when wrongly initialized)
-    return torch.nn.functional.relu(w - w_min) * eta_minus
-
-
-def _A_minus_hard(w, w_min, eta_minus):
-    """Hard lower bound for multiplicative STDP
-
+    def decay(
+        self,
+        z_pre,
+        z_post,
+        tau_pre_inv,
+        tau_post_inv,
+        a_pre,
+        a_post,
+        dt,
+    ):
+        self.t_pre += dt * tau_pre_inv * (-self.t_pre + a_pre * z_pre)
+        self.t_post += dt * tau_post_inv * (-self.t_post + a_post * z_post)
+
+
+# STDP parameters class
+class STDPParameters:
+    """STDP parameters.
     Parameters:
-        w (torch.Tensor): synaptic weight
-        w_min (float): lower bound on synaptic weight (should be < w_max)
-        eta_minus (float): contribution of post-pre activity to synaptic weight (should be >= 0)
+        a_pre (float): Contribution of presynaptic spikes to trace
+        a_post (float): Contribution of postsynaptic spikes to trace
+        tau_pre_inv (float): Inverse decay time constant of presynaptic spike trace in 1/s
+        tau_post_inv (float): Inverse decay time constant of postsynaptic spike trace in 1/s
+        w_min (float): Lower bound on synaptic weights (should be < w_max)
+        w_max (float): Upper bound on synaptic weight (should be > w_min)
+        eta_plus (float): Learning rate for synaptic potentiation (0 < eta_plus << 1)
+        eta_minus (float): Learning rate for synaptic depression (0 < eta_minus << 1)
+        stdp_algorithm (string): Algorithm for STDP updates. Options in {"additive","additive_step","multiplicative_pow","multiplicative_relu"}
+        mu (float): Exponent for multiplicative STDP (0 <= mu <= 1)
+        hardbound (boolean): Impose hardbounds by clipping (recommended unles eta_* << 1)
+        convolutional (boolean): Convolutional weighting kernel
+        stride (int): Stride for convolution
+        padding (int): Padding for convolution
+        dilation (int): Dilation for convolution
     """
-    return heaviside(w - w_min) * eta_minus
-
 
-def linear_soft_multiplicative_stdp_step(
-    w: torch.Tensor,
+    def __init__(
+        self,
+        a_pre=torch.as_tensor(1.0),
+        a_post=torch.as_tensor(1.0),
+        tau_pre_inv=torch.as_tensor(1.0 / 50e-3),
+        tau_post_inv=torch.as_tensor(1.0 / 50e-3),
+        w_min=0.0,
+        w_max=1.0,
+        eta_plus=1e-3,
+        eta_minus=1e-3,
+        stdp_algorithm="additive",
+        mu=0.0,
+        hardbound=True,
+        convolutional=False,
+        stride=1,
+        padding=0,
+        dilation=1,
+    ):
+        self.a_pre = a_pre
+        self.a_post = a_post
+        self.tau_pre_inv = tau_pre_inv
+        self.tau_post_inv = tau_post_inv
+        self.w_min = w_min
+        self.w_max = w_max
+        self.eta_plus = eta_plus
+        self.eta_minus = eta_minus
+        self.stdp_algorithm = stdp_algorithm
+
+        if self.stdp_algorithm == "additive":
```

Comment:
As far as I understood, by using callable we will be forcing the user to link the right initialisation function to the stepper. I think this will make the STDP functionality less plug and play.

Reply:
As far as I understood, by using callable we will be forcing the user to link the right initialisation function to the stepper. I think this will make the STDP functionality less plug and play.

Reply:
Just to make sure I understand @ChauhanT, would you say that it's too limiting to force the user to choose initialisation function in the STDPParameters case?

The current code looks conceptually pretty nice to me. I could still imagine factoring out the `stdp_algorithm` parameter to a function, so one would write, for instance `STDPParameters(..., Additive, ...)` instead of `STDPParameters(..., "additive", ...)`. But I'm also happy to leave it as-is :-)

Reply:
@Jegp I see what you mean now, and I think from the code-base point of view you are absolutely right. Splitting it into small initialisation functions is probably better. I hadn't understood what 'callable' means in python. I'll look into this next week ;)

Reply:
@Jegp  I have been giving this some thought. The initialisations in the if...else use various combinations of the member attributes for different STDP algorithms. To factor these out will mean creating module-scope callables which have a uniform interface. I think it will make the code a little more modular, but less readable. 

If the idea is to encourage users to implement their own STDP algorithms, they will want to write much more than only the initialisation functions for the parameters (such users are likely to adopt a different parametrisation of the STDP dynamics altogether). 

So I would vote for leaving the if else cascade in there for now.